src/Namecheap/Provider.php

```diff
@@ -0,0 +1,502 @@
+            NamecheapApi::CONTACT_TYPE_ADMIN      => $params->admin->register,
+            NamecheapApi::CONTACT_TYPE_TECH       => $params->tech->register,
+            NamecheapApi::CONTACT_TYPE_BILLING    => $params->billing->register,
+        ];
+
+        try {
+            $this->api()->register(
+                $domainName,
+                intval($params->renew_years),
+                $contacts,
+                $this->prepareNameservers($params, 'nameservers.ns'),
+            );
+
+            return $this->_getInfo($domainName, sprintf('Domain %s was registered successfully!', $domainName));
+        } catch (\Throwable $e) {
+            $this->handleException($e, $params);
+        }
+    }
+
+    /**
+     * @param  RegisterDomainParams  $params
+     *
+     * @return void
+     */
+    private function checkRegisterParams(RegisterDomainParams $params): void
+    {
+        if (!Arr::has($params, 'registrant.register')) {
+            throw $this->errorResult('Registrant contact data is required!');
+        }
+
+        if (!Arr::has($params, 'tech.register')) {
+            throw $this->errorResult('Tech contact data is required!');
+        }
+
+        if (!Arr::has($params, 'admin.register')) {
+            throw $this->errorResult('Admin contact data is required!');
+        }
+
+        if (!Arr::has($params, 'billing.register')) {
+            throw $this->errorResult('Billing contact data is required!');
+        }
+    }
+
+
+    /**
+     * @param ArrayAccess $params
+     * @param string $prefix
+     * @return string
+     */
+    private function prepareNameservers(ArrayAccess $params, string $prefix): string
+    {
+        $nameServers = "";
+
+        $custom = 0;
+        $default = 0;
+
+        for ($i = 1; $i <= self::MAX_CUSTOM_NAMESERVERS; $i++) {
+            if (Arr::has($params, $prefix.$i)) {
+                $host = Arr::get($params, $prefix.$i)->host;
+                if (!in_array($host, self::DEFAULT_NAMESERVERS)) {
+                    $nameServers .= $host.",";
+                    $custom++;
+                } else {
+                    $default++;
+                }
+            }
+        }
+
+        if ($custom != 0 && $default != 0) {
+            throw $this->errorResult(
+                "It's not possible to mix Namecheap's default nameservers with other ones",
+                $params
+            );
+        }
+
+        if ($custom + $default < self::MIN_CUSTOM_NAMESERVERS) {
+            throw $this->errorResult('Minimum two nameservers are required!', $params);
+        }
+
+        return $nameServers;
+    }
+
+    /**
+     * @param  TransferParams  $params
+     *
+     * @return DomainResult
+     */
+    public function transfer(TransferParams $params): DomainResult
+    {
+        $sld = Utils::normalizeSld($params->sld);
+        $tld = Utils::normalizeTld($params->tld);
+
+        if (!in_array($tld, NamecheapApi::TRANSFER_TLD)) {
```

Comment:
Let's not hardcode transferrable TLDs here as this may depend on the account-holder's agreement with Namecheap and may be subject to change over time. It's fine to just let Namecheap return an error response from their API instead.